### PR TITLE
Update Django to 1.11 and improve test setup

### PR DESCRIPTION
**Pull request by jeancochrane**
# Overview

This PR updates Django to version 1.11, the only version of Django 1.x that [has long-term support](https://www.djangoproject.com/download/#supported-versions). It also updates the test environment to provide development scripts and run tests in a matrix with tox, meaning that we can test all supported versions of Python/Django locally as well as on Travis.

## Notes

- As part of updating Django, a number of dependencies needed to be updated as well, including django-filter and django-rest-framework. While I had initially hoped to provide legacy support for Django 1.8, these dependencies no longer support Django <1.11, making this our floor version by default.

- During the update process I noticed that djsonb doesn't support Django 2.0+. As such, this PR is limited in scope to upgrading the 1.x line, and I've filed the work to support Django 2.0+ in its own issue (https://github.com/azavea/ashlar-2018-fellowship/issues/27).

## Testing

Travis will run tests for all supported versions of Python. To run these tests locally, use the `test` script:

```
./scripts/test
```

By default, tox will create a local cache of virtualenvs to make future testing easier. To free up disk space after testing, use the `clean` script: 

```
./scripts/clean
```

Closes https://github.com/azavea/ashlar-2018-fellowship/issues/10.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -1,21 +1,23 @@
 language: python
 python:
-- '2.7'
-- '3.3'
-- '3.4'
-- '3.5'
-- '3.6'
+  - '2.7'
+  - '3.4'
```
Comment (jeancochrane) [inline .travis.yml:4]: I dropped Python 3.3 as [this version of Python has reached end-of-life](https://www.python.org/downloads/release/python-337/).

File: ashlar/filters.py [hunk 2]
```diff
@@ -15,14 +16,26 @@
...
+        'filter_class': django_filters.CharFilter
+    }
+}
 
 class RecordFilter(GeoFilterSet):
 
-    record_type = django_filters.MethodFilter(name='record_type', action='filter_record_type')
-    polygon = django_filters.MethodFilter(name='polygon', action='filter_polygon')
-    polygon_id = django_filters.MethodFilter(name='polygon_id', action='filter_polygon_id')
+    record_type = django_filters.Filter(field_name='record_type', method='filter_record_type')
```
Comment (jeancochrane) [inline ashlar/filters.py:34]: All changes to this file are API changes to support `django-filter` v1.0+.

File: ashlar/pagination.py [hunk 3]
```diff
@@ -13,7 +13,7 @@ def paginate_queryset(self, queryset, request, view=None):
             return None
 
         self.offset = self.get_offset(request)
-        self.count = _get_count(queryset)
+        self.count = self.get_count(queryset)
```
Comment (jeancochrane) [inline ashlar/pagination.py:16]: `_get_count` became a method on the `LimitOffsetPagination` class in DRF 3.8.

File: ashlar/serializers.py [hunk 4]
```diff
@@ -128,6 +128,7 @@ def get_light_value(self, forecast_data, occurred_from):
 
     class Meta:
         model = Record
+        fields = '__all__'
```
Comment (jeancochrane) [inline ashlar/serializers.py:131]: Required API change for DRF 3.8

File: tests/test_date_filter.py [hunk 5]
```diff
@@ -74,19 +74,19 @@ def test_valid_datefilter(self):
         """ Test filtering on dates """
         self.assertEqual(len(self.queryset), 1)
 
-        req1 = Request(self.factory.get('/foo/', {'occurred_max': self.a_date}))
+        req1 = self.factory.get('/foo/', {'occurred_max': self.a_date})
```
Comment (jeancochrane) [inline tests/test_date_filter.py:77]: Changes to these tests all reflect the API changes in DRF 3.8 and django-filter 1.0+ that were implemented in the app code above.

File: Dockerfile [hunk 6]
```diff
@@ -1,26 +1,70 @@
...
+# Install Python versions for testing
+RUN pyenv install 2.7.15 && \
+    pyenv install 3.3.7 && \
+    pyenv install 3.4.8 && \
+    pyenv install 3.5.5 && \
+    pyenv install 3.6.6
+
+RUN pyenv global 3.6.6
+RUN pip install tox tox-pyenv
+RUN pyenv local 2.7.15 3.3.7 3.4.8 3.5.5 3.6.6
```
Comment (flibbertigibbet) [inline Dockerfile]: Could we parameterize these version numbers?

File: run_tests.py [hunk 7]
```diff
@@ -17,6 +17,10 @@ def create_extension_postgis():
 
 if __name__ == '__main__':
     from django.core.management import execute_from_command_line
+
+    if django.VERSION >= (1, 11):
+        django.setup()
```
Comment (flibbertigibbet) [inline run_tests.py]: Should this maybe yell at the user if the Django version is unsupported?

**Review by flibbertigibbet — APPROVED** [review]
:+1:

File: Dockerfile (hunk 6 above)
Comment (jeancochrane) [reply Dockerfile]: Great idea!

**Pull request merged**